bundler/lib/bundler/dsl.rb

```diff
@@ -447,6 +447,15 @@ def check_path_source_safety
     end
 
     def check_rubygems_source_safety
+      unless @sources.implicit_global_source?
+        Bundler::SharedHelpers.major_deprecation(2, <<-W.strip)
+Using Gemfile without a global source is deprecated
+and it will result in an error in the future Bundler versions.
+To fix this issue, simply define a global source in your Gemfile like this:
+source 'https://rubygems.org'
```

Comment:
Let's use the same string concatenation style used in the warnings below, so that we can keep the message nicely intented and automatically wrapped by the shell.

Regarding the wording, I'd go with

> This Gemfile does not include an explicit global source. Not using an explicit global source may result in a different lockfile being generated depending on the gems you have installed locally before bundler is run. Instead, define a global source in your Gemfile like this: `source "https://rubygems.org"`.

Finally, you can early return after printing the deprecation/error because the case is mutually exclusive with the multiple global sources case.

Reply:
I was thinking about extracting the `check_rubygems_source_safety` logic into two methods.

- `check_implicit_global_source`
- `check_multiple_global_source`

What do you think?

Reply:
Yeah, it makes sense! :+1:

Reply:
I ended up separating the warnings part only :D you will see.